# Convolutional Neural Network for CIFAR 10

In [ ]:
'''Train a simple deep CNN on the CIFAR10 small images dataset.
GPU run command:
    THEANO_FLAGS=mode=FAST_RUN,device=gpu,floatX=float32 python cifar10_cnn.py
It gets down to 0.65 test logloss in 25 epochs, and down to 0.55 after 50 epochs.
(it's still underfitting at that point, though).
Note: the data was pickled with Python 2, and some encoding issues might prevent you
from loading it in Python 3. You might have to load it in Python 2,
save it in a different format, load it in Python 3 and repickle it.
'''

from __future__ import print_function
from keras.datasets import cifar10
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Flatten
from keras.layers import Convolution2D, MaxPooling2D
from keras.optimizers import SGD
from keras.utils import np_utils


In [2]:
batch_size = 32
nb_classes = 10
nb_epoch = 200
data_augmentation = True

# input image dimensions
img_rows, img_cols = 32, 32
# the CIFAR10 images are RGB
img_channels = 3

# the data, shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = cifar10.load_data()
print('X_train shape:', X_train.shape)
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

X_train shape: (50000, 32, 32, 3)
50000 train samples
10000 test samples


In [3]:
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)

model = Sequential()
model.add(Convolution2D(32, 3, 3, border_mode='same',
                        input_shape=X_train.shape[1:]))
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
#model.add(Dropout(0.25))

model.add(Convolution2D(64, 3, 3, border_mode='same'))
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
#model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
#model.add(Dropout(0.5))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))

In [4]:
# let's train the model using SGD + momentum (how original).
sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy',
              optimizer=sgd,
              metrics=['accuracy'])

X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255


In [ ]:
if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(X_train, Y_train,
              batch_size=batch_size,
              nb_epoch=nb_epoch,
              validation_data=(X_test, Y_test),
              shuffle=True)
else:
    print('Using real-time data augmentation.')

    # this will do preprocessing and realtime data augmentation
    datagen = ImageDataGenerator(
        featurewise_center=False,  # set input mean to 0 over the dataset
        samplewise_center=False,  # set each sample mean to 0
        featurewise_std_normalization=False,  # divide inputs by std of the dataset
        samplewise_std_normalization=False,  # divide each input by its std
        zca_whitening=False,  # apply ZCA whitening
        rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
        horizontal_flip=True,  # randomly flip images
        vertical_flip=False)  # randomly flip images

    # compute quantities required for featurewise normalization
    # (std, mean, and principal components if ZCA whitening is applied)
    datagen.fit(X_train)

    # fit the model on the batches generated by datagen.flow()
    model.fit_generator(datagen.flow(X_train, Y_train,
                        batch_size=batch_size),
                        samples_per_epoch=X_train.shape[0],
nb_epoch=nb_epoch,
                        validation_data=(X_test, Y_test))

Using real-time data augmentation.
Epoch 1/200
50000/50000 [==============================] - 304s - loss: 2.0097 - acc: 0.2469 - val_loss: 1.6309 - val_acc: 0.4107
Epoch 2/200
50000/50000 [==============================] - 313s - loss: 1.4853 - acc: 0.4607 - val_loss: 1.2480 - val_acc: 0.5433
Epoch 3/200
50000/50000 [==============================] - 316s - loss: 1.2215 - acc: 0.5622 - val_loss: 1.0454 - val_acc: 0.6335
Epoch 4/200
50000/50000 [==============================] - 334s - loss: 1.0249 - acc: 0.6381 - val_loss: 0.9013 - val_acc: 0.6925
Epoch 5/200
50000/50000 [==============================] - 324s - loss: 0.9078 - acc: 0.6807 - val_loss: 0.8650 - val_acc: 0.7040
Epoch 6/200
50000/50000 [==============================] - 312s - loss: 0.8334 - acc: 0.7112 - val_loss: 0.8517 - val_acc: 0.7124
Epoch 7/200
50000/50000 [==============================] - 306s - loss: 0.7762 - acc: 0.7298 - val_loss: 0.7247 - val_acc: 0.7520
Epoch 8/200
50000/50000 [==============================

Following is the performance of the CNN algorithm -


Epoch 1/200
50000/50000 [==============================] - 304s - loss: 2.0097 - acc: 0.2469 - val_loss: 1.6309 - val_acc: 0.4107
Epoch 2/200
50000/50000 [==============================] - 313s - loss: 1.4853 - acc: 0.4607 - val_loss: 1.2480 - val_acc: 0.5433
Epoch 3/200
50000/50000 [==============================] - 316s - loss: 1.2215 - acc: 0.5622 - val_loss: 1.0454 - val_acc: 0.6335
Epoch 4/200
50000/50000 [==============================] - 334s - loss: 1.0249 - acc: 0.6381 - val_loss: 0.9013 - val_acc: 0.6925
Epoch 5/200
50000/50000 [==============================] - 324s - loss: 0.9078 - acc: 0.6807 - val_loss: 0.8650 - val_acc: 0.7040
Epoch 6/200
50000/50000 [==============================] - 312s - loss: 0.8334 - acc: 0.7112 - val_loss: 0.8517 - val_acc: 0.7124
Epoch 7/200
50000/50000 [==============================] - 306s - loss: 0.7762 - acc: 0.7298 - val_loss: 0.7247 - val_acc: 0.7520
Epoch 8/200
50000/50000 [==============================] - 345s - loss: 0.7340 - acc: 0.7447 - val_loss: 0.8117 - val_acc: 0.7290
Epoch 9/200
50000/50000 [==============================] - 345s - loss: 0.7008 - acc: 0.7567 - val_loss: 0.7216 - val_acc: 0.7549
Epoch 10/200
50000/50000 [==============================] - 331s - loss: 0.6650 - acc: 0.7684 - val_loss: 0.6674 - val_acc: 0.7712
Epoch 11/200
50000/50000 [==============================] - 326s - loss: 0.6448 - acc: 0.7764 - val_loss: 0.6361 - val_acc: 0.7826
Epoch 12/200
50000/50000 [==============================] - 317s - loss: 0.6167 - acc: 0.7873 - val_loss: 0.6233 - val_acc: 0.7903
Epoch 13/200
50000/50000 [==============================] - 321s - loss: 0.5960 - acc: 0.7941 - val_loss: 0.6418 - val_acc: 0.7852
Epoch 14/200
50000/50000 [==============================] - 341s - loss: 0.5745 - acc: 0.7996 - val_loss: 0.6184 - val_acc: 0.7928
Epoch 15/200
50000/50000 [==============================] - 329s - loss: 0.5562 - acc: 0.8080 - val_loss: 0.6104 - val_acc: 0.8014
Epoch 16/200
50000/50000 [==============================] - 319s - loss: 0.5564 - acc: 0.8065 - val_loss: 0.6346 - val_acc: 0.7932
Epoch 17/200
50000/50000 [==============================] - 313s - loss: 0.5420 - acc: 0.8099 - val_loss: 0.6113 - val_acc: 0.7993
Epoch 18/200
50000/50000 [==============================] - 314s - loss: 0.5231 - acc: 0.8181 - val_loss: 0.6304 - val_acc: 0.7879
Epoch 19/200
50000/50000 [==============================] - 314s - loss: 0.5097 - acc: 0.8228 - val_loss: 0.5800 - val_acc: 0.8070
Epoch 20/200
50000/50000 [==============================] - 329s - loss: 0.5053 - acc: 0.8245 - val_loss: 0.6007 - val_acc: 0.8061
Epoch 21/200
50000/50000 [==============================] - 347s - loss: 0.4967 - acc: 0.8281 - val_loss: 0.5936 - val_acc: 0.8049
Epoch 22/200
50000/50000 [==============================] - 322s - loss: 0.4868 - acc: 0.8305 - val_loss: 0.6188 - val_acc: 0.7997
Epoch 23/200
50000/50000 [==============================] - 323s - loss: 0.4750 - acc: 0.8351 - val_loss: 0.6131 - val_acc: 0.8084
Epoch 24/200
50000/50000 [==============================] - 299s - loss: 0.4680 - acc: 0.8386 - val_loss: 0.5882 - val_acc: 0.8111
Epoch 25/200
50000/50000 [==============================] - 323s - loss: 0.4639 - acc: 0.8392 - val_loss: 0.6057 - val_acc: 0.8127
Epoch 26/200
50000/50000 [==============================] - 329s - loss: 0.4574 - acc: 0.8413 - val_loss: 0.5957 - val_acc: 0.8100
Epoch 27/200
50000/50000 [==============================] - 325s - loss: 0.4487 - acc: 0.8459 - val_loss: 0.5757 - val_acc: 0.8109
Epoch 28/200
50000/50000 [==============================] - 321s - loss: 0.4425 - acc: 0.8478 - val_loss: 0.5974 - val_acc: 0.8088
Epoch 29/200
50000/50000 [==============================] - 334s - loss: 0.4420 - acc: 0.8495 - val_loss: 0.6082 - val_acc: 0.8009
Epoch 30/200
 3232/50000 [>.............................] - ETA: 298s - loss: 0.4005 - acc: 0.8645